library/payments.ts

```diff
@@ -0,0 +1,283 @@
+        },
+        ["id"]
+      )
+      .then((data) => {
+        if (!data[0]?.id) {
+          log.error(
+            `Error inserting new payment record into transaction table: ${data}`
+          );
+          reject();
+        }
+        resolve(data[0]?.id);
+      })
+      .catch((err) => {
+        log.debug(
+          `Error inserting new payment record into transaction table: ${err}`
+        );
+        reject();
+      });
+  });
+}
+
+async function handleCompletedPayment(
+  res: any, // TODO: Use express response type
+  userId: string,
+  msatAmount: number,
+  paymentRecordId: number,
+  paymentData: sendPaymentResponse
+) {
+  const trx = await db.knex.transaction();
+  // Update transaction table and user balance in one tx
+  return trx("transaction")
+    .update({
+      fee_msat: paymentData.data.fee,
+      payment_request: paymentData.data.invoice,
+      success: true,
+      is_pending: false,
+      external_id: paymentData.data.id,
+      preimage: paymentData.data.preimage,
+    })
+    .where({ id: paymentRecordId })
+    .then(() => {
+      // Decrement balance and unlock user
+      return trx("user")
+        .decrement({
+          msat_balance: msatAmount + parseInt(paymentData.data.fee),
+        })
+        .update({ updated_at: db.knex.fn.now(), is_locked: false })
+        .where({ id: userId });
+    })
+    .then(trx.commit)
+    .then(() => {
+      log.debug(`Payment successful for ${userId}`);
+      return res
+        ? res
+            .status(200)
+            .send({ invoiceId: paymentRecordId, ...paymentData.data })
+        : {
+            success: true,
+            data: {
+              invoiceId: paymentRecordId,
+              preimage: paymentData.data.preimage,
+              feeMsat: paymentData.data.fee,
+            },
+          };
+    })
+    .catch((err) => {
+      log.error(
+        `Error updating transaction table on handleCompletedPayment: ${err}`
+      );
+      return res
+        ? res.status(500).send("Payment succeeded but update failed")
+        : { success: false, error: "UNKNOWN" };
+    });
+}
+
+async function handleFailedPayment(
+  res: any, // TODO: Use express response type
+  userId: string,
+  paymentRecordId: number,
+  paymentData: sendPaymentResponse
+) {
+  const trx = await db.knex.transaction();
+  // Update transaction table and user status
+  return trx("transaction")
+    .update({
+      success: false,
+      is_pending: false,
+      failure_reason: paymentData.message,
+    })
+    .where({ id: paymentRecordId })
+    .then(() => {
+      // Unlock user
+      return trx("user")
+        .update({ updated_at: db.knex.fn.now(), is_locked: false })
+        .where({ id: userId });
+    })
+    .then(trx.commit)
+    .then(() => {
+      log.debug(`Payment failed for ${userId}`);
+      return res
+        ? res.status(400).send(`Payment failed: ${paymentData.message}`)
+        : { success: false, error: paymentData.message };
```

Comment:
looks like this ternary is driven by the argument `res`, should we instead have an early return if `res` is not defined?
```suggestion
async function handleFailedPayment(
  res: any, // TODO: Use express response type
  userId: string,
  paymentRecordId: number,
  paymentData: sendPaymentResponse
) {
  if (!res) return { success: false, error: paymentData.message };
  const trx = await db.knex.transaction();
  // Update transaction table and user status
  return trx("transaction")
    .update({
      success: false,
      is_pending: false,
      failure_reason: paymentData.message,
    })
    .where({ id: paymentRecordId })
    .then(() => {
      // Unlock user
      return trx("user")
        .update({ updated_at: db.knex.fn.now(), is_locked: false })
        .where({ id: userId });
    })
    .then(trx.commit)
    .then(() => {
      log.debug(`Payment failed for ${userId}`);
      return res.status(400).send(`Payment failed: ${paymentData.message}`)
```